### PR TITLE
hw23-ez

### PR DESCRIPTION
**Pull request by ellaz16**

### CODE REVIEW

File: homework/hw2-goals-approaches/ez/hw2-ez.md [hunk 1]
```diff
@@ -0,0 +1,390 @@
...
+# Prepare data
+df_notna = df.dropna(subset=["flavor_text", "release_date"])
+docs = df_notna.flavor_text.to_list()
+timestamps = df_notna.release_date.to_list()
+```
+
+```{code-cell} ipython3
+# Train a BERTopic model
+topic_model = BERTopic(verbose=True)
+topics, probs = topic_model.fit_transform(docs)
```
Comment (rtbs-dev) [inline homework/hw2-goals-approaches/ez/hw2-ez.md]: Is it necessary to retrain this? Can `ft_model` be used? This adds a lot to the time to run your notebook, and the work appears to already be done in the code, right?

**Review by rtbs-dev — COMMENTED** [review]
Before the final submission, it helps to add either an export of the notebook (html, pdf, etc) to version control (or, for extra fanciness, make them artifacts of a git hook or github workflow :) if you're using jupytext, or just add the ipynb for reports like this that don't hold a ton of state. That way I can see preliminary results before merging!

File: homework/hw2-goals-approaches/ez/hw2-ez.md [hunk 2]
```diff
@@ -0,0 +1,390 @@
...
+
+```{code-cell} ipython3
+# Get topic information
+ft_model.get_topic_info()[2:10].set_index("Topic")
+```
+
+#### Renaming Topics
+Topic 1: kami <br>
+Topic 2: goblin <br>
+Topic 3: wildspeaker <br>
```
Comment (rtbs-dev) [inline homework/hw2-goals-approaches/ez/hw2-ez.md]: Great insight. Would love to see a brief bullet list of why each of these makes sense (like a shorter version of your great Kami discussion, e.g. one bullet per topic, very short)

File: homework/hw2-goals-approaches/ez/hw2-ez.md [hunk 3]
```diff
@@ -0,0 +1,390 @@
...
+
+![topics_over_time.png](attachment:topics_over_time.png)
+
++++
+
+Kami is the Japanese word for "great spirits." From the figure, we can see that the kami topic peaks around 2004, suggesting that there was probably an expansion set about kami released in 2004. After some research, I found out that *Champions of Kamigawa* was released in October 2004 as the first set in the Kamigawa block and it introduced many rare creatures with kami-like magics. I actually started playing mtg recently, and bought the *Kamigawa: Neon Dynasty* expansion set (released in February 2022) over the weekend. According to the storyline, this set represents the current era on Kamigawa, which is more than 1200 years after conclusion of the original Kamigawa block. If we excluded the 2004 kami outlier, we would probably see another smaller spike in 2022.
+
+```{code-cell} ipython3
+# Visualize topics over time without the 'kami' outlier
+topic_model.visualize_topics_over_time(topics_over_time, topics=[0,1,3,4,5,6,7])
```
Comment (rtbs-dev) [inline homework/hw2-goals-approaches/ez/hw2-ez.md]: This may be my side only, but locally I'm seeing `kami` as the 0th topic, so this command doesn't remove it.

File: homework/hw2-goals-approaches/ez/hw2-ez.md [hunk 4]
```diff
@@ -0,0 +1,390 @@
...
+# Train-test split
+X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=111)
+
+# Fit OneVsRestClassifier
+multilabel_model.fit(X_train, y_train)
+```
+
+```{code-cell} ipython3
+# Save the model using pickle
+pickle.dump(multilabel_model, open("multilabel.sav", 'wb'))
```
Comment (rtbs-dev) [inline homework/hw2-goals-approaches/ez/hw2-ez.md]: Similar to the topic model, is this necessary? Should live in the source files, with this notebook only "reading" it, right? Did something not work from the pickle?

File: homework/hw2-goals-approaches/ez/multilabel.py [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+df = pd.read_feather("../../../data/mtg.feather")
+
+# X
+tfidf = TfidfVectorizer(
+    min_df=5, # ignore rare words (appear in less than 5 documents)
+    max_df=0.8, # ignore common words (appear in more than 80% of documents)
+    stop_words="english"
+)
+text = df["text"] + df["flavor_text"].fillna('')
+X = tfidf.fit_transform(text)
```
Comment (rtbs-dev) [inline homework/hw2-goals-approaches/ez/multilabel.py:24]: Ahhh, ok so I see the issue. This is what we use [`sklearn.pipeline.Pipeline`](https://scikit-learn.org/stable/modules/compose.html#pipeline) for: combine all preprocessing and transforming into one model, that you can save. Then you can use a consistent input through the whole notebook.

File: homework/hw2-goals-approaches/ez/multilabel.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+
+# y
+ci = df["color_identity"]
+cv = CountVectorizer(tokenizer=lambda x: x, lowercase=False)
+y = cv.fit_transform(ci)
```
Comment (rtbs-dev) [inline homework/hw2-goals-approaches/ez/multilabel.py:29]: Generally I see [`MultiLabelBinarizer`](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MultiLabelBinarizer.html) used for this, any advantage/reason for a coutvec here instead? Generally we do not allow counts more than 1 in multilabel predictions.

**Comment by rtbs-dev** [comment]
I've added some a commit with some minor edits to illustrate a couple things. Great work!
